### DISCUSSION TITLE
Maintain tags for something like @latest-2

### DISCUSSION THREAD

**Discussion by jdschleicher** [forcedotcom/cli#2853] [Ideas]
Hi! 

We have been using @latest in our CI server for quite some time. Every once in a while we get an install from latest that bites us. 

We are looking to implement some functionality that captures the latest 3 wednesday releases and using the 3rd latest version to be our standard cli install version.  

This may give us a some wiggle room in the event there is a bug introduced. 

It would be nice if we didn't have to build this logic and we could do something like :  **npm install sf@latest-3**

I'm not sure if this a helpful approach or not or if this tagging syntax doesn't make sense for salesforce cli. 

In my initial googl'ing this is all I came across to make sense of what I was looking for: https://stackoverflow.com/questions/40398488/how-does-npm-install-npmlatest-2-resolve-package-version

**Comment by mshanemc** [comment]
> Every once in a while we get an install from latest that bites us

Aww, sorry! [and you have opened issues for those, right ??]

We have the same problem with some of our dependencies in CI 😄. 

ex: our node version is almost always the default, but we can set an env when we need something in particular.
https://github.com/salesforcecli/github-workflows/blob/deba43938b2a907e476b897fcf90424f49b06da2/.github/workflows/unitTestsLinux.yml#L13

Maybe `@latest` could be in the env, but then in an emergency you change it to `@2.40.7` etc.  (you can use the [RN](https://github.com/forcedotcom/cli/blob/main/releasenotes/README.md) for the list of "Wednesday" versions).  And don't forget to put it back when you're satisfied.

The one downside to your idea (ex: `-3`) is that it would update every week.  

The upside to parameterizing it is that you could also use `nightly` or `latest-rc` on a cron job to know if the upcoming `latest` will be a problem for your project.  Then you can 
1. report it so we fix it **before** it becomes latest
2. switch the env to a version you know works ok before the one that doesn't becomes `latest`

**Reply by jdschleicher** [reply]
All good! we will go down the path of vars in gh actions ( basically have this done :) ) 

the issues that we had were reported but I don't think were directly related to the cli update but what the cli consumes? 

Probably not saying that right.  We are in govcloud so we also dont get updates the same as everyone else.  Here is an example issue we made:  https://github.com/forcedotcom/cli/issues/2413

**Reply by RupertBarrow** [reply]
@jdschleicher 
I always pin versions to make the pipeline stable.
Why not update manually to @latest-3 every month, or so ?

**Reply by jdschleicher** [reply]
@mshanemc are there any plans that would adjust the format of the release notes page? 

There looks to be a expected pattern that we can capture to get stable release notes from here: https://raw.githubusercontent.com/forcedotcom/cli/main/releasenotes/README.md

We could call this directly and parse the third latest stable version? Maybe use that as a cron job and set that as the version to be the parameter? Overkill?

@RupertBarrow we could definitely do that, trying to automate it for now but it wouldn't be too crazy of a lift to do manually

**Reply by mshanemc** [reply]
>@mshanemc are there any plans that would adjust the format of the release notes page?

yes.  I wouldn't consider that a stable API.

there's a "index" json file for each architecture (they should all have the same versions, but with different links, so any of them would do for you) 
https://developer.salesforce.com/docs/atlas.en-us.sfdx_setup.meta/sfdx_setup/sfdx_setup_install_cli.htm#sfdx_setup_install_cli_olderversions

those are stable APIs (and what we use for `sf update --interactive` and `sf update --available`

you'd want the highest patch version of [how many versions you want to go back] (highest minor is a nightly, followed by rc, then stable, stable-1, etc)

**Reply by jdschleicher** [reply]
awesome thank you!

**Reply by mshanemc** [reply]
we've also recently started uploading release artifacts (tarballs, signed installers) to the github releases 
ex: https://github.com/salesforcecli/cli/releases/tag/2.46.6

**Reply by jdschleicher** [reply]
thank you @mshanemc !
